eos-update-flatpak-repos

```diff
@@ -574,26 +610,63 @@ def _migrate_installed_flatpaks():
             logging.info('Found {}/{}/{} to migrate'
                          .format(name, arch, branch))
 
+            old_ostree_ref = '{}:{}'.format(ref.get_origin(), ref.format_ref())
+
             try:
                 new_branch = migrate.get('new-branch')
                 if new_branch:
                     _update_branch(ref, new_branch, refs)
+                    branch = new_branch
+                    ref = _refresh_ref(inst, kind, name, arch, branch)
 
                 deploy_dir = ref.get_deploy_dir()
                 deploy = os.path.join(deploy_dir, 'deploy')
 
                 new_origin = migrate.get('new-origin')
                 if new_origin:
                     _update_deploy_file_with_origin(deploy, new_origin)
+                    origin = new_origin
+                    ref = _refresh_ref(inst, kind, name, arch, branch)
 
                 new_subpaths = migrate.get('new-subpaths')
                 if new_subpaths:
                     _update_deploy_file_with_subpaths(deploy, new_subpaths)
+                    subpaths = new_subpaths
+                    ref = _refresh_ref(inst, kind, name, arch, branch)
+
+                new_ostree_ref = '{}:{}'.format(ref.get_origin(), ref.format_ref())
+
+                # in the case we have only modified the subpaths, the ref does
+                # not change, so we should remove it from the list of old refs
+                # which will be deleted.
+                if new_ostree_ref == old_ostree_ref:
+                    logging.debug('Ostree ref {} unchanged'.format(old_ostree_ref))
+                    matching_ostree_refs.remove(old_ostree_ref)
+                    continue
+
+                # when the branch or origin has changed, we copy the old ref to
+                # the new one, pointing at the same commit
+                if old_ostree_ref in ostree_refs and \
+                   not new_ostree_ref in ostree_refs:
+                    logging.debug('Copying ref {} to {}'.format(old_ostree_ref,
+                                                                new_ostree_ref))
+                    repo.set_ref_immediate(ref.get_origin(), ref.format_ref(),
+                                           ostree_refs[old_ostree_ref])
+                    ostree_refs[new_ostree_ref] = ostree_refs[old_ostree_ref]
+
             except Exception:
                 logging.exception('Failure applying migration to {}'
                                   .format(name))
                 continue
 
+        # any refs we copied, along with any matches for apps we migrated
+        # in the past, should be deleted.
+        for old_ostree_ref in matching_ostree_refs:
+            logging.debug('Deleting old/stale ref {}'.format(old_ostree_ref))
```

Comment:
Maybe log this as a message, rather than a debug message, and also include the checksum of the ref (before it’s deleted), or the new name of the ref, so that we have some scope for manually undoing any erroneous deletions if they happen in future (and we have access to the log output from this script).